extensions/controllers/sandboxclaim_controller.go

```diff
@@ -341,3 +364,103 @@ func (r *SandboxClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Owns(&sandboxv1alpha1.Sandbox{}).
 		Complete(r)
 }
+
+// reconcileNetworkPolicy ensures a NetworkPolicy exists for the claimed Sandbox.
+func (r *SandboxClaimReconciler) reconcileNetworkPolicy(ctx context.Context, claim *extensionsv1alpha1.SandboxClaim, template *extensionsv1alpha1.SandboxTemplate) error {
+	logger := log.FromContext(ctx)
+
+	// 1. Cleanup Check: If disabled/missing, delete existing policy
+	if template == nil || template.Spec.NetworkPolicy == nil || !template.Spec.NetworkPolicy.Enabled {
+		existingNP := &networkingv1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      claim.Name + "-network-policy",
+				Namespace: claim.Namespace,
+			},
+		}
+		if err := r.Delete(ctx, existingNP); err != nil {
+			if !k8errors.IsNotFound(err) {
+				logger.Error(err, "Failed to clean up disabled NetworkPolicy")
+				return err
+			}
+		} else {
+			logger.Info("Deleted disabled NetworkPolicy", "name", existingNP.Name)
+		}
+		return nil
+	}
+
+	np := &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      claim.Name + "-network-policy",
+			Namespace: claim.Namespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.Client, np, func() error {
+		nameHash := sandboxcontrollers.NameHash(claim.Name)
+
+		// 2. Target the Sandbox Pod
+		np.Spec.PodSelector = metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				sandboxLabel: nameHash,
+			},
+		}
+		np.Spec.PolicyTypes = []networkingv1.PolicyType{
+			networkingv1.PolicyTypeIngress,
+			networkingv1.PolicyTypeEgress,
+		}
+
+		// 3. Ingress Rules
+		var ingressRules []networkingv1.NetworkPolicyIngressRule
+		templateNP := template.Spec.NetworkPolicy
+
+		// A. Helper: IngressControllerSelectors
+		if sel := templateNP.IngressControllerSelectors; sel != nil {
+			peer := networkingv1.NetworkPolicyPeer{
+				NamespaceSelector: sel.NamespaceSelector,
+				PodSelector:       sel.PodSelector,
+			}
+			ingressRules = append(ingressRules, networkingv1.NetworkPolicyIngressRule{
+				From: []networkingv1.NetworkPolicyPeer{peer},
+			})
+		}
+
+		// B. Standard Rules
+		if len(templateNP.AdditionalIngressRules) > 0 {
+			ingressRules = append(ingressRules, templateNP.AdditionalIngressRules...)
+		}
+
+		np.Spec.Ingress = ingressRules
+
+		// 4. Egress Rules
+		var egressRules []networkingv1.NetworkPolicyEgressRule
+
+		// A. Default DNS Rule (Port 53 Allow-All)
+		// We keep this 'secure-by-default' behavior
```

Comment:
This adds more allow rules, and conflates with this comment about secure-by-default. Would you clarify the reason of hardcoding DNS here? Also, this prevents users from creating strictly isolated sandboxes, and won't support environments where DNS runs on non-standard ports. If that's intentional, please document it.